### PR TITLE
Add ServiceIndex component

### PR DESCRIPTION
**Pull request by caseydavenport**
## Description

This PR:
- Adds a new ServiceIndex component which handles generating IP sets from Kubernetes EndpointSlices for use with service-based policy rules.
- Adds a new field to the proto.Rule structure to track IP+port type IP sets, which match on both IP and port.

It relies on two other PRs in the series:
- New API: https://github.com/projectcalico/api/pull/8/
- Handling in libcalico-go: https://github.com/projectcalico/libcalico-go/pull/1443

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests
- [ ] eBPF data plane changes
- [ ] Windows data plane changes

## Release Note

```release-note
Felix support for matching Kubernetes services in egress rules.
```

### CODE REVIEW

**Comment by caseydavenport** [comment]
>   It looks like that the field DstIpPortSetIds is a new addition to the proto.Rule struct. Please check if it affects XDP optimizations in any way, update the isValidRuleForXDP function and the "invalid policies" unit test if needed, and then add the name to the knownProtoRuleFields slice above. Please DO NOT blindly add the field to the slice without checking its influence on the XDP optimization.

I know what the test output tells me to do... but I'm wondering if maybe I should just blindly add the field to the slice without checking its influence on the XDP optimization? :thinking:

File: serviceindex/serviceindex.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+	}
+	return &idx
+}
+
+func (idx *ServiceIndex) RegisterWith(allUpdDispatcher *dispatcher.Dispatcher) {
+	allUpdDispatcher.Register(model.ResourceKey{}, idx.OnUpdate)
+}
+
+// OnUpdate makes ServiceIndex compatible with the Dispatcher.  It accepts
+// updates for endpoints and profiles and passes them through to the Update/DeleteXXX methods.
```
Comment (fasaxc) [inline serviceindex/serviceindex.go:67]: Copy-pasta

File: serviceindex/serviceindex.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+func (idx *ServiceIndex) UpdateEndpointSlice(es *discovery.EndpointSlice) {
+	svc := serviceName(es)
+	if _, ok := idx.endpointSlicesByService[svc]; !ok {
+		idx.endpointSlicesByService[svc] = map[string]*discovery.EndpointSlice{}
+	}
+	k := fmt.Sprintf("%s/%s", es.Namespace, es.Name)
+	cached := idx.endpointSlices[k]
+	oldIPSetContributions := idx.membersFromEndpointSlice(cached)
```
Comment (fasaxc) [inline serviceindex/serviceindex.go:92]: Push this expensive op inside the active check below?

File: serviceindex/serviceindex.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	if !ok {
+		return
+	}
+
+	// Determine the service that contributed this endpoint slice.
+	svc := serviceName(es)
+	if ipSet, ok := idx.activeIPSetsByService[svc]; ok {
+		// Active service has had an EndpointSlice deleted. Iterate all the ip set members
+		// contributed by this endpoint slice and decref them. For those which go from 1 to 0,
+		// we should end a membership removal from the data plane.
```
Comment (fasaxc) [inline serviceindex/serviceindex.go:151]: typo: end

File: serviceindex/serviceindex.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+func (idx *ServiceIndex) membersFromEndpointSlice(es *discovery.EndpointSlice) []labelindex.IPSetMember {
+	if es == nil {
+		// A nil endpoint slice produces no IP set members.
+		return nil
+	}
+
+	// Create a member for each endpoint + port combination. If there
+	// are no ports specified, it means no ports (thus, not IP set membership). If nil is specified,
```
Comment (fasaxc) [inline serviceindex/serviceindex.go:187]: typo: not

File: serviceindex/serviceindex.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+		logc.Debugf("New IP set")
+	} else if curr.ServiceName == serviceName {
+		// Not a new IP set - we're already tracking it as an active service-based
+		// IP set, so nothing to do.
+		logc.Debugf("Known IP set, nothing to do")
+		return
+	} else {
+		// This branch means that a new service name has generated the same ID as another.
+		// This branch is not possible because the ID is uniquely generated from the service name.
+		panic(fmt.Sprintf("BUG: Same ID generated for two service names: %s and %s", curr.ServiceName, serviceName))
```
Comment (fasaxc) [inline serviceindex/serviceindex.go:239]: Use `logrus.Panic` because it flushes the log on the way down.

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looking great.  Think there are still a few TODOs:

- UTs (including some calc graph FVs for the new service index)
- FVs: should be able to mark those with BPF-SAFE too so that they run in both suites.
- Windows

File: dataplane/windows/policysets/policysets.go [hunk 6]
```diff
@@ -438,6 +439,8 @@ func (s *PolicySets) protoRuleToHnsRules(policyId string, pRule *proto.Rule, isI
 		}
 	}
 
+	// TODO: Implement match on IP+port IP sets.
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply dataplane/windows/policysets/policysets.go]: Leaving out for now to get this in, but will submit a follow-up shortly after.

**Comment by caseydavenport** [comment]
@fasaxc I believe I've got all the necessary tests covered now! Could you give this a second look?

File: fv/service_policy_test.go [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+			},
+			{
+				Action: api.Allow,
+			},
+		}
+		_, err = client.NetworkPolicies().Create(utils.Ctx, allowServicePolicy, utils.NoOptions)
+		Expect(err).NotTo(HaveOccurred())
+
+		// Expect traffic is denied to the endpoint specified in the service - w1, TCP 80
+		// Traffic the other direciton, and to other ports should be allowed.
```
Comment (fasaxc) [inline fv/service_policy_test.go:268]: Typo: direciton

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looking good; just a couple of points:
* I do like to have the odd test in the calc graph FV framework since they do cover a lot of the invariants of the calc graph that are tricky to test elsewhere.
* As you flagged, CI is failing.

**Pull request merged**